### PR TITLE
Map stores guide

### PR DESCRIPTION
**Pull request by SirCipher**
Closes #194

### CODE REVIEW

File: src/_rust-server/map-stores.md [hunk 1]
```diff
@@ -0,0 +1,395 @@
...
+                context.update(ExampleAgent::STORE, key, i32::max(state, new_value))
+            })
+    }
+}
+
+```
+
+# Bounds and Initialisation
+
+Map Stores place similar bounds to the standard library's `HashMap` but with the addition of requiring both the key and value types to implement the `swimos_form::Form` and `Send` traits. When the store is first intiialised, the store is initialised to an empty map. If you need to populate it with another value, then use the agent `on_start` lifecycle event handler and populate the inital value of the store there, ensuring that you handle agent restarts.
```
Comment (horned-sphere) [inline src/_rust-server/map-stores.md:74]: Might want to point out that initializing in `on_start` will go badly if you have persistence on. We should probably add a way to do one time only iniitialization of lanes/stores.

File: src/_rust-server/map-stores.md [hunk 2]
```diff
@@ -0,0 +1,395 @@
...
+
+# Use cases
+
+Map Stores are useful in situtations where you need to persist data that is derived from lane events received but you do not wish for it to be available outside of the scope of the agent. While it is possible to keep state in the agent's lifecycle, this is lost when the agent is stopped and must be rebuilt each time the agent is started. Using a Map Store, the state is retreived from the underlying persistence engine when the agent first starts.
```
Comment (horned-sphere) [inline src/_rust-server/map-stores.md:78]: Ther are other advantages of stores:

- Values are mutable without having to faff around with `RefCell`s in lifecycles.
- They can be accessed by event handlers that get suspended in futures (these can't access the lifecycle).

File: src/_rust-server/map-stores.md [hunk 3]
```diff
@@ -71,11 +71,11 @@ impl ExampleLifecycle {
 
 # Bounds and Initialisation
 
-Map Stores place similar bounds to the standard library's `HashMap` but with the addition of requiring both the key and value types to implement the `swimos_form::Form` and `Send` traits. When the store is first intiialised, the store is initialised to an empty map. If you need to populate it with another value, then use the agent `on_start` lifecycle event handler and populate the inital value of the store there, ensuring that you handle agent restarts.
+Map Stores place similar bounds to the standard library's `HashMap` but with the addition of requiring both the key and value types to implement the `swimos_form::Form` and `Send` traits. When the store is first intiialised, the store is initialised to an empty map. If the state of the store is to be populated using the `on_start` handler, then care must be taken to not overwrite any previously held state, as it is not currently possible to detect restarts; a workaround is to wrap the state in an `Option` and only replace it if it is `None`.
```
Comment (horned-sphere) [inline src/_rust-server/map-stores.md:74]: Actually, probably just adding a `initialized: ValueStore<bool>` would be better as then you don't have to keep checking the `Option` every time you use it.

File: src/_rust-server/map-stores.md [hunk 4]
```diff
@@ -12,15 +12,15 @@ redirect_from:
 
 # Overview
 
-A Map Store provides an agent with a store for map state that is not publically addressable. If server [persistence]({% link _rust-server/persistence.md %}) is enabled, then the state of the store will be persisted to the underlying engine and after a server restart, any state will be restored. If you require addressable scalar state, then a [Value Lane]({% link _rust-server/value-lanes.md %}) will be suitable.
+A Map Store provides an agent with a store for map state that is not publically addressable. If server [persistence]({% link _rust-server/persistence.md %}) is enabled, then the state of the store will be persisted to the underlying engine and after a server restart, any state will be restored. If you require addressable value state, then a [Value Lane]({% link _rust-server/value-lanes.md %}) will be suitable.
```
Comment (horned-sphere) [inline src/_rust-server/map-stores.md:15]: I'm stil not happy with the server restart bit.

Comment (SirCipher) [reply src/_rust-server/map-stores.md]: I've removed it. Rather than explaining it here just for stores, I'll cover it in more detail in an agent guide

**Review by horned-sphere — APPROVED** [review]

**Pull request merged**